File: src/text_renderer.zig
const std = @import("std");
const zm = @import("zmath");
const c = @import("c.zig");
const Shader = @import("shader.zig").Shader;
const log = std.log;
const Allocator = std.mem.Allocator;
const AutoHashMap = std.AutoHashMap;

const Character = struct {
    texture: u32,
    size: zm.Vec,
    bearing: zm.Vec,
    advance: u64,
};

pub const TextRenderer = struct {
    const Self = @This();
    characters: AutoHashMap(u8, Character),
    shader: Shader,
    vao: u32,
    vbo: u32,

    pub fn init(allocator: *Allocator) !Self {
        var characters = try createCharacters(allocator);
        var shader = Shader.init();
        shader.create(allocator, .vertex, "resources/shaders/text.vert");
        shader.create(allocator, .fragment, "resources/shaders/text.frag");
        shader.createProgram();

        shader.use();
        var projection = zm.orthographicRh(800.0, 600.0, -10.0, 10.0);
        shader.set_mat4("projection", projection);

        var VAO: u32 = undefined;
        var VBO: u32 = undefined;

        c.glGenVertexArrays(1, &VAO);
        c.glGenBuffers(1, &VBO);
        c.glBindVertexArray(VAO);
        c.glBindBuffer(c.GL_ARRAY_BUFFER, VBO);
        c.glBufferData(c.GL_ARRAY_BUFFER, @sizeOf(f32) * 6 * 4, null, c.GL_DYNAMIC_DRAW);
        c.glEnableVertexAttribArray(0);
        c.glVertexAttribPointer(0, 4, c.GL_FLOAT, c.GL_FALSE, 4 * @sizeOf(f32), null);
        c.glBindBuffer(c.GL_ARRAY_BUFFER, 0);
        c.glBindVertexArray(0);

        return Self{
            .characters = characters,
            .shader = shader,
            .vao = VAO,
            .vbo = VBO,
        };
    }

    pub fn deinit(self: *Self) void {
        self.characters.deinit();
        self.shader.deinit();
    }

    pub fn draw(self: *Self, text: []const u8, x: f32, y: f32, scale: f32, color: zm.Vec) void {
        // activate corresponding render state
        var x_tmp = x;
        self.shader.use();
        self.shader.set_vec3("textColor", color);

        c.glActiveTexture(c.GL_TEXTURE0);
        c.glBindVertexArray(self.vao);

        for (text) |char| {
            var nullableCharacter = self.characters.get(char);
            var character: Character = undefined;
            if (nullableCharacter) |value| {
                character = value;
            } else {
                character = self.characters.get('$').?;
            }

            var xpos = x_tmp + character.bearing[0] * scale;
            var ypos = y - (character.size[1] - character.bearing[1]) * scale;

            var w = character.size[0] * scale;
            var h = character.size[1] * scale;
            // update VBO for each character
            var vertices = &[_][4]f32{
                [_]f32{ xpos, ypos + h, 0.0, 0.0 },
                [_]f32{ xpos, ypos, 0.0, 1.0 },
                [_]f32{ xpos + w, ypos, 1.0, 1.0 },
                [_]f32{ xpos, ypos + h, 0.0, 0.0 },
                [_]f32{ xpos + w, ypos, 1.0, 1.0 },
                [_]f32{ xpos + w, ypos + h, 1.0, 0.0 },
            };
            // render glyph texture over quad
            c.glBindTexture(c.GL_TEXTURE_2D, character.texture);
            // update content of VBO memory
            c.glBindBuffer(c.GL_ARRAY_BUFFER, self.vbo);
            c.glBufferSubData(c.GL_ARRAY_BUFFER, 0, @sizeOf(f32) * 6 * 4, vertices); // be sure to use glBufferSubData and not glBufferData

            c.glBindBuffer(c.GL_ARRAY_BUFFER, 0);
            // render quad
            c.glDrawArrays(c.GL_TRIANGLES, 0, 6);
            // now advance cursors for next glyph (note that advance is number of 1/64 pixels)
            x_tmp += @intToFloat(f32, character.advance >> 6) * scale; // bitshift by 6 to get value in pixels (2^6 = 64 (divide amount of 1/64th pixels by 64 to get amount of pixels))
        }
        c.glBindVertexArray(0);
        c.glBindTexture(c.GL_TEXTURE_2D, 0);
    }

    fn createCharacters(allocator: *Allocator) !AutoHashMap(u8, Character) {
        var characters = AutoHashMap(u8, Character).init(allocator.*);
        var ft: c.FT_Library = undefined;

        if (c.FT_Init_FreeType(&ft) != 0) {
            @panic("FREETYPE: Could not init FreeType Library");
        }

        var path = "resources/fonts/DejaVuSans.ttf";

        var face: c.FT_Face = undefined;
        if (c.FT_New_Face(ft, path, 0, &face) != 0) {
            @panic("FREETYPE: Failed to load font");
        } else {
            _ = c.FT_Set_Pixel_Sizes(face, 0, 48);
            c.glPixelStorei(c.GL_UNPACK_ALIGNMENT, 1);

            var char: u8 = 0;
            while (char < 128) : (char += 1) {
                if (c.FT_Load_Char(face, char, c.FT_LOAD_RENDER) != 0) {
                    log.err("FREETYTPE: Failed to load Glyph", .{});
                    continue;
                }

                var texture: u32 = undefined;
                c.glGenTextures(1, &texture);
                c.glBindTexture(c.GL_TEXTURE_2D, texture);
                c.glTexImage2D(
                    c.GL_TEXTURE_2D,
                    0,
                    c.GL_RED,
                    @intCast(i32, face.*.glyph.*.bitmap.width),
                    @intCast(i32, face.*.glyph.*.bitmap.rows),
                    0,
                    c.GL_RED,
                    c.GL_UNSIGNED_BYTE,
                    face.*.glyph.*.bitmap.buffer,
                );

                c.glTexParameteri(c.GL_TEXTURE_2D, c.GL_TEXTURE_WRAP_S, c.GL_CLAMP_TO_EDGE);
                c.glTexParameteri(c.GL_TEXTURE_2D, c.GL_TEXTURE_WRAP_T, c.GL_CLAMP_TO_EDGE);
                c.glTexParameteri(c.GL_TEXTURE_2D, c.GL_TEXTURE_MIN_FILTER, c.GL_LINEAR);
                c.glTexParameteri(c.GL_TEXTURE_2D, c.GL_TEXTURE_MAG_FILTER, c.GL_LINEAR);

                var character = Character{
                    .texture = texture,
                    .size = zm.f32x4(
                        @intToFloat(f32, face.*.glyph.*.bitmap.width),
                        @intToFloat(f32, face.*.glyph.*.bitmap.rows),
                        0.0,
                        0.0,
                    ),
                    .bearing = zm.f32x4(
                        @intToFloat(f32, face.*.glyph.*.bitmap_left),
                        @intToFloat(f32, face.*.glyph.*.bitmap_top),
                        0.0,
                        0.0,
                    ),
                    .advance = @intCast(u64, face.*.glyph.*.advance.x),
                };
                try characters.put(char, character);
            }
            c.glBindTexture(c.GL_TEXTURE_2D, 0);
        }

        _ = c.FT_Done_Face(face);
        _ = c.FT_Done_FreeType(ft);

        return characters;
    }
};

File: src/text_renderer.cpp
#include <stdio.h>
#include "text_renderer.h"

#include <ft2build.h>
#include FT_FREETYPE_H

int TextRenderer::create_characters() {
    FT_Library ft;

    if (FT_Init_FreeType(&ft)) {
        printf("ERROR::FREETYPE: Could not init FreeType Library\n");
        return -1;
    }

    const char *font_name = (char *)"assets/fonts/Antonio-Bold.ttf";

    FT_Face face;
    if (FT_New_Face(ft, font_name, 0, &face)) {
        printf("ERROR::FREETYPE: Failed to load font\n");
        return -1;
    }

	FT_Set_Pixel_Sizes(face, 0, bitmap_font_size);

	// disable byte-alignment restriction
	glPixelStorei(GL_UNPACK_ALIGNMENT, 1);

	// load first 128 characters of ASCII
	for (unsigned char c = 0; c < 128; c++) {
		if (FT_Load_Char(face, c, FT_LOAD_RENDER)) {
			printf("ERROR::FREETYTPE: Failed to load Glyph\n");
			continue;
		}
		unsigned int texture;
		glGenTextures(1, &texture);
		glBindTexture(GL_TEXTURE_2D, texture);
		glTexImage2D(GL_TEXTURE_2D,
					 0,
					 GL_RED,
					 face->glyph->bitmap.width,
					 face->glyph->bitmap.rows,
					 0,
					 GL_RED,
					 GL_UNSIGNED_BYTE,
					 face->glyph->bitmap.buffer
					 );

		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

		glm::vec2 size;
		size[0] = face->glyph->bitmap.width;
		size[1] = face->glyph->bitmap.rows;
		glm::vec2 bearing;
		bearing[0] = face->glyph->bitmap_left;
		bearing[1] = face->glyph->bitmap_top;

		Character character = {
			texture,
			glm::ivec2(face->glyph->bitmap.width, face->glyph->bitmap.rows),
			glm::ivec2(face->glyph->bitmap_left, face->glyph->bitmap_top),
			static_cast<unsigned int>(face->glyph->advance.x)
		};
		characters[c] = character;

	}
	glBindTexture(GL_TEXTURE_2D, 0);
    FT_Done_Face(face);
    FT_Done_FreeType(ft);

	return 0;
}

TextRenderer::TextRenderer(glm::mat4 &projection) {
	bitmap_font_size = 100.0f;

	shader = new Shader();
	shader->create(GL_VERTEX_SHADER, (char *)"shaders/text.vert");
	shader->create(GL_FRAGMENT_SHADER, (char *)"shaders/text.frag");
	shader->create_program();

	shader->use();
	shader->set_mat4((char *)"projection", projection);

	create_characters();

	unsigned int VAO, VBO;
    glGenVertexArrays(1, &VAO);
    glGenBuffers(1, &VBO);
    glBindVertexArray(VAO);
    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBufferData(GL_ARRAY_BUFFER, sizeof(float) * 6 * 4, NULL, GL_DYNAMIC_DRAW);
    glEnableVertexAttribArray(0);
    glVertexAttribPointer(0, 4, GL_FLOAT, GL_FALSE, 4 * sizeof(float), 0);
    glBindBuffer(GL_ARRAY_BUFFER, 0);
    glBindVertexArray(0);

	vao = VAO;
	vbo = VBO;
}

TextRenderer::~TextRenderer() {
	delete shader;
}

void TextRenderer::draw(std::string text, glm::vec2 position, float font_size, glm::vec3 color) {
    // activate corresponding render state
	auto x = position.x;
	auto y = position.y;
	auto scale = font_size / bitmap_font_size;

    shader->use();
	shader->set_vec3((char *)"textColor", color);

    glActiveTexture(GL_TEXTURE0);
    glBindVertexArray(vao);

	//auto offset = calculate_text_offset(text, scale, alignment);
	auto offset = glm::vec2(0.0f);
    // iterate through all characters
    for (long unsigned int i = 0; i != text.size(); i++) {
		char c = text[i];
        Character ch = characters[(int)c];

        float xpos = x + offset.x + ch.bearing.x * scale;
        float ypos = y + offset.y + (ch.size.y - ch.bearing.y) * scale;

        float w = ch.size.x * scale;
        float h = ch.size.y * scale;

        // update VBO for each character
        float vertices[6][4] = {
            { xpos,     ypos - h,   0.0f, 0.0f },
            { xpos,     ypos,       0.0f, 1.0f },
            { xpos + w, ypos,       1.0f, 1.0f },

            { xpos,     ypos - h,   0.0f, 0.0f },
            { xpos + w, ypos,       1.0f, 1.0f },
            { xpos + w, ypos - h,   1.0f, 0.0f }
        };

        glBindTexture(GL_TEXTURE_2D, ch.texture);
        glBindBuffer(GL_ARRAY_BUFFER, vbo);
        glBufferSubData(GL_ARRAY_BUFFER, 0, sizeof(vertices), vertices);

        glBindBuffer(GL_ARRAY_BUFFER, 0);
        glDrawArrays(GL_TRIANGLES, 0, 6);

        x += (ch.advance >> 6) * scale;
    }

    glBindVertexArray(0);
    glBindTexture(GL_TEXTURE_2D, 0);
}

//glm::vec2 TextRenderer::calculate_text_offset(std::string text, float scale, TextAlignment alignment) {
//	auto offset = glm::vec2(0.0f);
//
//	if (alignment == TextAlignment::Left)
//		return glm::vec2(0.0f);
//
//	for (auto c : text) {
//        Character ch = characters[(int)c];
//
//		offset.x -= ch.size.x * scale;
//	}
//	offset.x = offset.x / 2.0f;
//
//	return offset;
//}
